Add current cuda device index to FXGraphCache key

jamesjwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147464

This PR intends to fix the cache related issues from https://github.com/pytorch/pytorch/issues/147405. 
It does *not* handle the dynamo recompile case in process, because it does not introduce any extra guards. For FXGraphCache and AOTAutogradCache, we simply have to have the device context in the cache key. 

Note that for any function that accepts tensor inputs, the device context is naturally already included in the cache key by the metadata of example inputs. However, for functions that return constants or have no arguments, the device context still needs to be in the cache key. 

A more robust fix for this would be to have inductor generate device guards that are dynamic, instead of specialized. This would also help us share more cache artifacts. 

I've added unit tests for FXGraphCache and AOTAutogradCache, both of which would fail without this change.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D69875939](https://our.internmc.facebook.com/intern/diff/D69875939)

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/codecache.py:
```diff
@@ -808,6 +808,11 @@ def __init__(
+        # This device index is usually already encoded by the device of the inputs
+        # but fx graphs don't necessarily have tensor inputs
+        if torch.cuda.is_available():
+            self.default_cuda_device_index = torch.cuda.current_device()
```
partially a PSA / question for @albanD  - to what extent do you expect people to be using `torch.accelerator.current_device()` + PT2 today? James and I talked about it offline a bit - we'll need to do something similar for other accelerators in the long run to avoid accelerator-specific warm cache problems

albanD replied:
Yes you should use torch.accelerator instead of cuda for these things. They lead to the same thing for the cuda device but you also get rocm/mtia/xpu support for free.

jamesjwu replied:
Is it guaranteed that:

if torch.cuda.is_available(), then `torch.accelerator.is_available()`? 
And that `torch.cuda.current_device() == torch.accelerator.current_device`? 

If so happy to change

jamesjwu replied:
Seems like it works in my tests, will run with it!

bdhirsh reviewed (approved):
sgtm! Although we should fix the dynamo guard issue too :)

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu commented:
@jamesjwu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69875939).

jamesjwu committed:
Update on "Add current cuda device index to FXGraphCache key"


This PR intends to fix the cache related issues from https://github.com/pytorch/pytorch/issues/147405. 
It does *not* handle the dynamo recompile case in process, because it does not introduce any extra guards. For FXGraphCache and AOTAutogradCache, we simply have to have the device context in the cache key. 

Note that for any function that accepts tensor inputs, the device context is naturally already included in the cache key by the metadata of example inputs. However, for functions that return constants or have no arguments, the device context still needs to be in the cache key. 

A more robust fix for this would be to have inductor generate device guards that are dynamic, instead of specialized. This would also help us share more cache artifacts. 

I've added unit tests for FXGraphCache and AOTAutogradCache, both of which would fail without this change.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D69875939](https://our.internmc.facebook.com/intern/diff/D69875939)

[ghstack-poisoned]

jamesjwu commented:
@jamesjwu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69875939).

jamesjwu commented:
Rebase to rerun tests
